Document deploying VEBA as K8s App (#116)
vmware-samples/vcenter-event-broker-appliance: vmware-event-router/hack/README.md

# Deploy vCenter Event Broker Application to existing Kubernetes Cluster

For customers with an existing Kubernetes Cluster, you can deploy the underlying components that make up the vCenter Event Broker Appliance. The instructions below will guide you in downloading the required files and using the `create_k8s_config.sh` shell script to aide in deploying the VEBA K8s application.

> The script will prompt users for the required input and  automatically setup and deploy both OpenFaaS and the VMware Event Router giving you a similar setup like the vCenter Event Broke Appliance.

Reviewer: Can we make OpenFaaS optional? I suspect some users will find the script useful but already have OpenFaaS installed/want to customize OpenFaaS (persistency, scalability, etc.). Also, this might keep the possibility to add AWS EventBridge later/manually.

Author: Going to keep the logic simple for now and assume OpenFaaS will be part of the deployment. In theory, L100-106 can simply be commented out to prevent the deployment of OpenFaaS. I'd like to get some feedback and we can always adjust

Author: Just added a new option in script to not deploy OpenFaaS

## Pre-Req:
* Deployment access to your K8s Cluster using kubectl
* Outbound connectivity or access to private  registry from the K8s Cluster to download the required containers to deploy OpenFaaS and VMware Event Router

## Install:

Step 1 - Clone the OpenFaaS to your local system

```
git clone https://github.com/openfaas/faas-netes
```

Step 2 - Change into the `faas-netes` directory and checkout version `0.9.2` which has been tested with VEBA

```
cd faas-netes

> git checkout 0.9.2

Reviewer: This (old version we use) is another reason for leaving the OpenFaaS installation to the user/optional.

```

> Step 3 - Download the `create_k8s_config.sh` script and ensure it has executable permission (`chmod +x create_k8s_config.sh`).

Reviewer: It would be good to specify that this script needs to be run from within the OpenFaaS git folder you check out above. Alternatively, you can use curl to retrieve this file and pipe it to a SHELL (which most installers pulling from Github do these days as well).

Author: Actually, the script runs outside of the OpenFaaS directory. Corrected the instructions and in case of debugging, I prefer the user has the script on hand

> Step 4 - Run the `create_k8s_config.sh` script which will prompt for vCenter Server address (FQDN/IP Address), the vCenter Server username and password which is authorized to retrieve vCenter Server Events (readOnly role is sufficient) and the admin password for OpenFaaS. Prior to deploying, you will be asked to confirm the input in case you need to change it.

Reviewer: Can be remove by comment above.

Author: N/A due to previous resolved comment

> For customers with an existing Kubernetes Cluster, you can deploy the underlying components that make up the vCenter Event Broker Appliance. The instructions below will guide you in downloading the required files and using the `create_k8s_config.sh` shell script to aide in deploying the VEBA K8s application.

Reviewer: ```suggestion
For customers with an existing Kubernetes ("K8s") cluster, you can deploy the underlying components that make up the vCenter Event Broker Appliance. The instructions below will guide you in downloading the required files and using the `create_k8s_config.sh` shell script to aide in deploying the VEBA K8s application.
```

> The script will prompt users for the required input and automatically setup and deploy both OpenFaaS and the VMware Event Router components giving you a similar setup like the vCenter Event Broke Appliance.. If you have already deployed OpenFaaS, you can skip that step during the script input.

Reviewer: ```suggestion
The script will prompt users for the required input and automatically setup and deploy both OpenFaaS and the VMware Event Router components giving you a similar setup like the vCenter Event Broke Appliance. If you have already deployed OpenFaaS, you can skip that step during the script input phase.
```

## Pre-Req:
* Ability to create namespaces, secrets and deployments in your K8s Cluster using kubectl

> * Outbound connectivity or access to private  registry from the K8s Cluster to download the required containers to deploy OpenFaaS and/or VMware Event Router

Reviewer: ```suggestion
* Outbound connectivity or access to private registry from the K8s Cluster to download the required containers to deploy OpenFaaS and/or VMware Event Router
```

Here is an example of what you should see if the deployment was successful:

![](example1.png)

> Step 5 - Ensure that all pods are running in both OpenFaaS and VMware namespace:

Reviewer: ```suggestion
Step 4 - Ensure that all pods are running in both OpenFaaS and VMware namespace:
```

> > **Note:** If your K8s cluster does not use Contour or support an HTTPS-based proxy, then the default OpenFaaS endpoint `http://<ip>:31112`

Reviewer: ```suggestion
> **Note:** If you don't use an Ingress controller, load-balancer or other means to expose your Kubernetes deployments (services), then the default OpenFaaS endpoint is `http://<worker-ip>:31112`
```

### Uninstall

To remove the VEBA and OpenFaaS K8s application, run the following commands:

```
kubectl delete ns vmware
kubectl delete -f faas-netes/yaml
kubectl delete -f faas-netes/namespaces.yml
```

## Deploy only VMware Event Router

Step 1 - Download the `create_k8s_config.sh` script and ensure it has executable permission (`chmod +x create_k8s_config.sh`).

Step 2 - Run the `create_k8s_config.sh` script which will prompt for vCenter Server address (FQDN/IP Address), the vCenter Server username and password which is authorized to retrieve vCenter Server Events (readOnly role is sufficient). Prior to deploying, you will be asked to confirm the input in case you need to change it.

```
./create_k8s_config.sh
```

Here is an example of what you should see if the deployment was successful:

![](example2.png)

> Step 5 - Ensure the VMware Event Router pod is running in the VMware namespace:

Reviewer: ```suggestion
Step 3 - Ensure the VMware Event Router pod is running in the VMware namespace:
```

```
# kubectl get pods -n vmware
NAME                                   READY   STATUS    RESTARTS   AGE
vmware-event-router-6744cc6447-xbpmn   1/1     Running   1          42s
```

## Uninstall:

> To remove the VEBA K8s application, run the following command:

Reviewer: ```suggestion
To remove the VMware Event Router K8s application, run the following command:
```
